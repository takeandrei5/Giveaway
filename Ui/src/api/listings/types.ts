import { PaginatedResult } from '../../utils/types';

export type ItemData = { id: string; image: string; title: string; createdAt: Date };

export type CreateListingRequest = {
	title: string;
	description: string;
	category: number;
	images: string[];
};

export type UpdateListingRequest = {
	title: string;
	description: string;
	category: number;
	images: string[];
};

export interface FetchListingsResponse {
	listings: PaginatedResult<ItemData>;
}

export type ListingInformation = {
	id: string;
	title: string;
	description: string;
	category: number;
	createdAt: Date;
	images: string[];
};

export type OwnerInformation = {
	email: string;
	name: string;
	image: string;
};

export type FetchListingDetailsResponse = { listingInfo: ListingInformation } & { ownerInfo: OwnerInformation };
