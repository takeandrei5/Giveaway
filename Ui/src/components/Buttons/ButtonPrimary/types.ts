import { ButtonProps } from '@chakra-ui/react';

export type ButtonPrimaryProps = Omit<ButtonProps, 'color'>;
