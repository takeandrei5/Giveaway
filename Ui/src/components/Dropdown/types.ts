export type DropdownProps = {
	id: string;
	options: DropdownOption[];
	onChangeHandler: (value: string) => void;
	name: string;
	value: string | number | readonly string[];
	isInvalid?: false;
};

export type DropdownOption = {
	value: string | number;
	displayValue: string;
};
