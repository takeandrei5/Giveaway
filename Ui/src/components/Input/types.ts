export type InputProps = {
	id: string;
	name: string;
	placeholder: string;
	value?: string;
	onBlur?: (e: React.ChangeEvent<HTMLInputElement> | React.ChangeEvent<HTMLTextAreaElement>) => void;
	onChange?: (e: React.ChangeEvent<HTMLInputElement> | React.ChangeEvent<HTMLTextAreaElement>) => void;
	onKeyDown?: (e: React.KeyboardEvent<HTMLInputElement> | React.KeyboardEvent<HTMLTextAreaElement>) => void;
	onKeyPress?: (e: React.KeyboardEvent<HTMLInputElement> | React.KeyboardEvent<HTMLTextAreaElement>) => void;
	onKeyUp?: (e: React.KeyboardEvent<HTMLInputElement> | React.KeyboardEvent<HTMLTextAreaElement>) => void;
	disabled?: boolean;
	multiline?: boolean;
	isInvalid?: boolean;
	height?: string | number;
	label?: string;
	leftIcon?: JSX.Element;
	rightIcon?: JSX.Element;
	rows?: number;
	width?: string | number;
};
