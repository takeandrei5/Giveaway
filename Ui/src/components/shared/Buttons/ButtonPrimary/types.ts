import { ButtonProps } from '@chakra-ui/react';

export type ButtonPrimaryProps = ButtonProps;
