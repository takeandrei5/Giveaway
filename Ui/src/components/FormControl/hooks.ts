import { FieldMetaProps } from 'formik';
import { useEffect, useState } from 'react';

import { useCheckFormIsInvalidProps } from './types';

const useCheckFormIsInvalid = (meta: FieldMetaProps<unknown>): useCheckFormIsInvalidProps => {
	const [isInvalid, setIsInvalid] = useState(false);

	useEffect(() => {
		setIsInvalid(meta.touched && !!meta.error);
	}, [meta.touched, meta.error]);

	return { isInvalid };
};

export { useCheckFormIsInvalid };
