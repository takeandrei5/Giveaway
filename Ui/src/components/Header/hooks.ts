import { NextRouter, useRouter } from 'next/router';
import { useCallback } from 'react';

const useLogin = () => {
	const router: NextRouter = useRouter();

	const handleSignIn: () => Promise<boolean> = useCallback(
		async () => await router.replace('/api/auth/login'),
		[router]
	);
	const handleSignInWithReturnTo: (returnTo: string) => Promise<boolean> = useCallback(
		async (returnTo: string) => await router.replace(`/api/auth/login?returnTo=${returnTo}`),
		[router]
	);
	const handleSignOut: () => Promise<boolean> = useCallback(() => router.replace('/api/auth/logout'), [router]);

	return { handleSignIn, handleSignInWithReturnTo, handleSignOut };
};

export { useLogin };
