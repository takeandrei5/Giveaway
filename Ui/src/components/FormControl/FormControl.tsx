import { FormControl as FormControlChakra, FormErrorMessage, FormLabel } from '@chakra-ui/react';
import { useField } from 'formik';
import React from 'react';

import { useCheckFormIsInvalid } from './hooks';
import { FormControlProps } from './types';

const FormControl = ({ children, id, label, name }: FormControlProps): JSX.Element => {
	const [_, meta] = useField<unknown>(name);
	const { isInvalid } = useCheckFormIsInvalid(meta);

	return (
		<FormControlChakra isInvalid={isInvalid} isRequired width='100%'>
			<FormLabel htmlFor={id} margin='0'>
				{label}
			</FormLabel>
			{React.cloneElement(children, { isInvalid })}
			{isInvalid && <FormErrorMessage marginTop='0.125rem'>{meta.error}</FormErrorMessage>}
		</FormControlChakra>
	);
};

export default FormControl;
