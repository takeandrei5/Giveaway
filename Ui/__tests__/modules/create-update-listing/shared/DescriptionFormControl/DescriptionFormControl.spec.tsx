import { useCheckFormIsInvalid } from '@components/FormControl/hooks';
import { DescriptionFormControl } from '@modules/create-update-listing/shared/DescriptionFormControl';
import { DescriptionFormControlProps } from '@modules/create-update-listing/shared/DescriptionFormControl/types';
import { render } from '@testing-library/react';
import { FieldMetaProps, useField } from 'formik';

jest.mock('formik', () => ({
	...jest.requireActual('formik'),
	useField: jest.fn(),
}));

jest.mock('@components/FormControl/hooks', () => ({
	useCheckFormIsInvalid: jest.fn((meta: FieldMetaProps<unknown>) => ({})),
}));

describe('DescriptionFormControl', () => {
	it('should match snapshot', () => {
		// Arrange
		const props: DescriptionFormControlProps = {
			name: 'test-name',
		};

		(useField as unknown as jest.Mock).mockImplementation(() => [
			{
				onBlur: jest.fn(),
				onChange: jest.fn(),
				value: 'test-name',
			},
		]);

		(useCheckFormIsInvalid as unknown as jest.Mock).mockImplementation((meta: FieldMetaProps<unknown>) => ({
			isInvalid: false,
		}));

		// Act
		const component = render(<DescriptionFormControl {...props} />);

		// Assert
		expect(component).toMatchSnapshot();
	});
});
