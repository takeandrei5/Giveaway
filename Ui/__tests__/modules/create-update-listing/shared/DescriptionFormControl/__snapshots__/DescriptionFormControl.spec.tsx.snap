// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DescriptionFormControl should match snapshot 1`] = `
<div>
  <div
    class="chakra-form-control css-8atqhb"
    role="group"
  >
    <label
      class="chakra-form__label css-xgxc5v"
      for="description-input"
      id="field-1-label"
    >
      Description:
      <span
        aria-hidden="true"
        class="chakra-form__required-indicator css-0"
        role="presentation"
      >
        *
      </span>
    </label>
    <textarea
      aria-required="true"
      class="chakra-textarea css-aky5ro"
      data-testid="textarea"
      id="description-input"
      name="test-name"
      placeholder="Be as detailed as you wish others were!"
      required=""
      rows="10"
    >
      test-name
    </textarea>
  </div>
</div>
`;
